Ctpdev: ctp aod readout

lietava opened the pull request.

lietava committed:
debug: aod ctpreadout

lietava committed:
clang

lietava commented:
HI @shahor02 , I am debugging 'ctpreadout' creation in AOD.
To generate data and run workflow I use macro derived from:
https://github.com/AliceO2Group/O2DPG/blob/master/MC/run/examples/O2DPG_pp_minbias_one_tf_qc.sh.
However it seems that trd trigger info:
https://github.com/lietava/AliceO2/blob/40293d7932b02075876abbf180f4f42404580cef/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L1181
is not available at:
https://github.com/lietava/AliceO2/blob/40293d7932b02075876abbf180f4f42404580cef/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L115
How can I get it ? Or should I use different container ?

shahor02 commented:
@lietava You can add TRD to https://github.com/AliceO2Group/AliceO2/blob/6cf03ec14f1b080df1351c7e21223444232af8be/Detectors/AOD/src/aod-producer-workflow.cxx#L55

lietava committed:
debug: trying to get TRD inputs

lietava commented:
Hi @shahor02 , I added TRD to the line you suggested. I also added TRD to the line 587 (o2-aod-producer-workflow) in 
[workflowrl.json.txt](https://github.com/AliceO2Group/AliceO2/files/9414568/workflowrl.json.txt).
I still get:
[626342:aod-producer-workflow]: I am here 01
[626342:aod-producer-workflow]: [10:41:11][ERROR] No TRD triggers
[626342:aod-producer-workflow]: I am here 1
R.

shahor02 commented:
This means that the workflow does not require TRD as an input source. I think you should add to 
https://github.com/AliceO2Group/O2DPG/blob/7def6eb42c8f883d3823f3adde22b019927b1437/MC/bin/o2dpg_sim_workflow.py#L1082

```
aodinfosources += ',TRD'
```

lietava commented:
@shahor02 @mpuccio : so in summary I added TRD (see above):
1.) aod-producer-worlflow.cxx
2.) workflowrl.json 
3.) o2dpg_sim_workflow.py
Still 
[636075:aod-producer-workflow]: [10:40:15][ERROR] No TRD triggers
What else ?
R.

shahor02 commented:
@lietava you did not need to touch the json file, it is anyway generated on the fly. Are you sure you have installed the O2DPG after modifying the `o2dpg_sim_workflow.py`? If so, could you put the tarball with your simulation on the cernbox?

lietava commented:
@shahor02 : I just modified o2dpg_sim_workflow.py in my /home/rl/O2DPG/MC/bin.
How I reinstalled O2DPG ? Recompile O2Physics ? R.

shahor02 commented:
If you are modifying master branch of O2DPG, it should be enough to run `alibuild build O2DPG`, otherwise, you should build whole O2PDPSuite. But, for the fast check, you can simply modify the `$O2DPG_ROOT/MC/bin/o2dpg_sim_workflow.py` (will be overwritten at the rebuild)

lietava commented:
"But, for the fast check, you can simply modify the $O2DPG_ROOT/MC/bin/o2dpg_sim_workflow.py (will be overwritten at the rebuild)" -- but this is what I just did ? R.

shahor02 commented:
Isn't ` /home/rl/O2DPG/MC/bin` your source directory? What is you $O2DPG_ROOT ?

lietava commented:
export O2DPG_ROOT=/home/rl/O2DPG

shahor02 commented:
$O2DPG_ROOT as well as other base dirs. are set by the alienv. Anyway, could you put one of your simulated TFs on the cernbox?

lietava commented:
@shahor02 please, see email

shahor02 commented:
OK, the TRD info is read but the AODProducer does not use it since it should be requested as a "cluster" rather than "track" type.
Please add:
```
if (src[GID::TRD]) {
   dataRequest->requestTRDTracklets(false);
}
```
at 
https://github.com/AliceO2Group/AliceO2/blob/ae867a6afd003a4fd9bc48794c38f6816d0d8fe8/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L2111

lietava committed:
fix: related to aod ctpreadout

lietava committed:
fix: cpmmenting cout

lietava committed:
clang

lietava commented:
@shahor02 @mpuccio : it seems it works now. Thanks. We can try on real data. But I think there is still at least one problem: access to CCDB does not use run number.

shahor02 commented on DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx:
```diff
@@ -1180,7 +1180,12 @@ gsl::span<const o2::trd::CalibratedTracklet> RecoContainer::getTRDCalibratedTrac
 {
-  return inputsTRD->mTriggerRecords;
+  if (inputsTRD == nullptr) {
+    LOG(ERROR) << "No TRD triggers";
```
if the TRD is not there, this is not an error, you can probably log a warning once, no need to do this at every TF.

lietava replied:
so I added warning counter to RecoContainr. h. OK ?

shahor02 replied:
yes, but it seems you did not push it.

lietava replied:
yes, I will, thanks.

lietava replied:
I pushed. Hmm there is problem - I try to increase counter in const function:
https://github.com/lietava/AliceO2/blob/a68ed6692d911e8da791b4f4751510500af938f2/DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx#L1186
Shall I give up this message ?
R.

shahor02 replied:
you can use local static counter.

lietava replied:
done

shahor02 commented on Detectors/AOD/src/AODProducerWorkflowSpec.cxx:
```diff
@@ -1662,6 +1667,7 @@ void AODProducerWorkflowDPL::run(ProcessingContext& pc)
   // helper map for fast search of a corresponding class mask for a bc
   std::unordered_map<uint64_t, uint64_t> bcToClassMask;
   if (mInputSources[GID::CTP]) {
+    LOG(info) << "CTP input available";
```
These logs at every TF is also too much, printing at 1st TF only should be enough.

lietava replied:
info changed to debug

shahor02 reviewed (commented):
Could you clarify where `access to CCDB does not use run number` problem is?

lietava commented:
> Could you clarify where `access to CCDB does not use run number` problem is?

it is line https://github.com/lietava/AliceO2/blob/db4affa28cbf0a6087f4598c2e86d27b59d81b9a/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L2094
where for MC I do not need specify run as validity is fir all MCs.
For real data It needs to be changed to something like:
ccdbParamSpec("CTP/Config/Config", bool runDependent =1, std::vector<CCDBMetadata> metadata = {{"runNumber","12345"}}, int64_t rate=0);
?
R.

shahor02 commented:
> it is line https://github.com/lietava/AliceO2/blob/db4affa28cbf0a6087f4598c2e86d27b59d81b9a/Detectors/AOD/src/AODProducerWorkflowSpec.cxx#L2094 where for MC I do not need specify run as validity is fir all MCs. For real data It needs to be changed to something like: ccdbParamSpec("CTP/Config/Config", bool runDependent =1, std::vector metadata = {{"runNumber","12345"}}, int64_t rate=0); ? R.

You just need to change it to 
```
dataRequest->inputs.emplace_back("ctpconfig", "CTP", "CTPCONFIG", 0, Lifetime::Condition, ccdbParamSpec("CTP/Config/Config", true));
```
But this will make it run dependent both for data and MC. Using `useMC` flag would not be reliable as we may run the AODproducer on MC run w/o this flag. So, we would either need an extra option telling that this is an MC data or to have per-run CTP objects also for CTP, as we have for e.g. GRPECS.

lietava committed:
fix: logs reduction

lietava committed:
clang

lietava committed:
fix: only first warning

lietava committed:
fix: def detector CTP

lietava committed:
dev: CTPconfig run dpendence as parameter

lietava committed:
clang

lietava commented:
@shahor02 : for CTP config I created parameter:
https://github.com/lietava/AliceO2/blob/0843fc6bb0f7aef7ae0ea79e12b2207b81a7e8a7/Detectors/AOD/src/aod-producer-workflow.cxx#L42

shahor02 commented:
note that your PR is still in draft mode some the CI is not running

lietava committed:
fix: print format

lietava committed:
fix: format

alibuild commented:
Error while checking build/O2/fullCI for 51be3396687ac4236d97a0d696dde7775c9a1c56 at 2022-08-30 17:34:
```
## sw/BUILD/o2checkcode-latest/log
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/Common/test/testGPUsortCUDA.cu:22:10: error: 'boost/test/unit_test.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUTracking/TRDTracking/GPUTRDTracker.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstruction.cxx:37:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUTracking/Base/GPUReconstructionCPU.cxx:45:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUTracking/Base/cuda/GPUReconstructionCUDAGenRTC.cu:16:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/cuda/../Shared/Utils.h:26:10: error: 'boost/program_options.hpp' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUTracking/display/GPUDisplay.cxx:36:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/Framework/Logger/include/Framework/Logger.h:14:10: error: 'fairlogger/Logger.h' file not found [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:520:12: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1059:69: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1175:5: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:1852:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:2276:18: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/cuda/std/detail/libcxx/include/type_traits:4336:16: error: CUDA device code does not support variadic functions [clang-diagnostic-error]
/usr/local/cuda-11.7/include/sm_20_atomic_functions.h:89:39: error: redefinition of 'atomicAdd' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:8: error: unknown type name '__host__' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:27: error: expected ';' after top level declarator [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:33: error: overloaded 'operator+=' must have at least one parameter of class or enumeration type [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:44: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:64:53: error: unknown type name 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:66:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:67:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:68:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:4: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:69:11: error: member reference base type 'int' is not a structure or union [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:85:1: error: unknown type name '__global__' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:19: error: use of undeclared identifier 'blockIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:32: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:45: error: use of undeclared identifier 'threadIdx' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/Kernels.hip.cxx:90:78: error: use of undeclared identifier 'blockDim' [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/../Shared/Utils.h:146:34: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/GPU/GPUbenchmark/hip/benchmark.hip.cxx:199:35: error: use of undeclared identifier 'int4'; did you mean 'int'? [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:33:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:33:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALCalibExtractor.h:33:10: error: 'omp.h' file not found [clang-diagnostic-error]
/sw/SOURCES/O2/9683-slc8_x86-64/0/Detectors/TOF/calibration/src/TOFChannelCalibrator.cxx:23:10: error: 'omp.h' file not found [clang-diagnostic-error]
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/9683/51be3396687ac4236d97a0d696dde7775c9a1c56/build_O2_fullCI/pretty.html).

lietava committed:
fix: format commented

lietava commented:
@shahor02 : can we merge ?
